runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,481 @@
+  if (dnode->numData) {
+      void **arr = dnode->data;
+      for (int i = 0; i < dnode->numData; i++) {
+          chpl_mem_free(arr[i], 0, 0);
+      }
+  }
+  chpl_mem_free(dnode->data, 0, 0);
+}
+
+// Initializes TLS; should only need to be called once.
+static void init_tls(void);
+static void init_tls(void) {
+  struct tls_node *node = chpl_calloc(1, sizeof(struct tls_node));
+  node->epoch = get_global_epoch();
+
+  // Append to head of list
+  struct tls_node *old_head;
+  do {
+    old_head = chpl_qsbr_tls_list;
+    node->next = old_head;
+  } while (!atomic_compare_exchange_weak_uintptr_t(
+      (uintptr_t *) &chpl_qsbr_tls_list, (uintptr_t) old_head, (uintptr_t) node)
+    );
+
+  CHPL_TLS_SET(chpl_qsbr_tls, node);
+}
+
+// Obtains the minimum epoch of all threads.
+static uint64_t safe_epoch(void);
+static uint64_t safe_epoch(void) {
+  uint64_t min = (uint64_t) -1;
+  // Check all remaining threads for whether they have updated to a more recent epoch.
+  for (struct tls_node *node = get_tls_list(); node != NULL; node = node->next) {
+    if (get_tasks(node) > 0) {
+      uint64_t epoch = get_epoch(node);
+      min = min ? (min < epoch) : epoch;       
+    }
+  }
+  
+  return min;
+}
+
+// Requires spinlock.
+static inline void handle_deferred_data(struct tls_node *node);
+static inline void handle_deferred_data(struct tls_node *node) {
+  // Acquire all data that can be deleted.
+  uint64_t epoch = safe_epoch();
+  struct defer_node *list = pop_bulk_defer_list(node, epoch);
+
+  // Handle deletion of said data...
+  for (struct defer_node *dnode = list; dnode != NULL;) {
+    struct defer_node *tmp = dnode;
+    dnode = dnode->next;
+    delete_data(tmp);
+    push_recycle_list(node, tmp);
+  }
+}
+
+void chpl_qsbr_init(void) {
+    CHPL_TLS_INIT(chpl_qsbr_tls);
+}
+
+void chpl_qsbr_onTaskCreation(void) {
+  struct tls_node *tls = CHPL_TLS_GET(chpl_qsbr_tls);
+  if (tls == NULL) {
+    init_tls();
+    tls = CHPL_TLS_GET(chpl_qsbr_tls);
+  }
+
+  // Notify that we are now in-use.
+  add_task(tls);
+}
+
+
+void chpl_qsbr_onTaskDestruction(void) {
+  struct tls_node *tls = CHPL_TLS_GET(chpl_qsbr_tls);
+  if (tls == NULL) {
+    init_tls();
+    tls = CHPL_TLS_GET(chpl_qsbr_tls);
+  }
+
+  // Notify that we have one less task
+  uint64_t nTasks = remove_task(tls) - 1;
+
+  // If we are the last task, handle any extra deferred data...
+  if (nTasks == 0) {
+    acquire_spinlock(tls);
+
+    // Empty as best we can
+    handle_deferred_data(tls);
+    if (tls->deferList == NULL) {
+      release_spinlock(tls);
+      return;
+    }
+
+    // At this point if we have any deferred data, we must pass this to some
+    // other thread, as we could be parked indefinitely. Find first active
+    // thread to dump the rest of our work on...
+    struct tls_node *node = get_tls_list();
+    for (; node != NULL; node = node->next) {
+      if (get_tasks(node) > 0) {
+        // Double check...
+        acquire_spinlock(node);
```

Comment:
I agree the double check is necessary here

Reply:
This reminded me that I do not check if `node == tls` and so we would end up deadlocking on `acquire_spinlock(node)` since we already acquired it above.